services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionAsyncClient.java

```diff
@@ -16,32 +16,79 @@
 package software.amazon.awssdk.services.s3.internal.crossregion;
 
 import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
 import java.util.function.Function;
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.awscore.AwsRequestOverrideConfiguration;
+import software.amazon.awssdk.endpoints.Endpoint;
 import software.amazon.awssdk.services.s3.DelegatingS3AsyncClient;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.endpoints.S3EndpointParams;
+import software.amazon.awssdk.services.s3.endpoints.S3EndpointProvider;
 import software.amazon.awssdk.services.s3.model.S3Request;
 
 @SdkInternalApi
 public final class S3CrossRegionAsyncClient extends DelegatingS3AsyncClient {
-
     public S3CrossRegionAsyncClient(S3AsyncClient s3Client) {
         super(s3Client);
     }
 
     @Override
-    protected <T extends S3Request, ReturnT> ReturnT invokeOperation(T request, Function<T, ReturnT> operation) {
-        Optional<String> bucket = request.getValueForField("bucket", String.class);
+    protected <T extends S3Request, ReturnT> CompletableFuture<ReturnT>
+            invokeOperation(T request, Function<T, CompletableFuture<ReturnT>> operation) {
+
+        Optional<String> bucket = request.getValueForField("Bucket", String.class);
 
         if (!bucket.isPresent()) {
             return operation.apply(request);
         }
 
-        //TODO: add modifyRequest logic
-        return operation.apply(request);
+        return operation.apply(requestWithDecoratedEndpointProvider(request, bucket.get()))
```

Comment:
What does `handleOperationFailure` do? Should we just return the future from `apply` instead?

```
CompletableFuture<T> result = operation.apply(...);
result.whenComplete((r, t) -> handleOperationFailure(...));
return result;
```

The reason for this is that if we return a new future (from `whenComplete`), then we have to handle the case where the caller cancels the future; however "real" S3 client should already have set that up the future returned from `apply`

Reply:
Another thought, if the `result` gets cancelled by the caller, then we should handle that case in `whenComplete` as well since we probably don't want to remove that cache entry in that case.

Reply:
Makes sense. Added this comment to the caching task description.